### PR TITLE
Adding flow diagrams for create/read resource, and some search components

### PR DESCRIPTION
**Pull request by YazanMSFT**
## Description
There are a lot of components involved when performing an operation on the FHIR server, such as inserting a resource or running a search. I thought it might be useful to include flow diagrams that show which components are called in what order in the happy path of some scenarios.

The flow diagrams are defined using [PlantUML](https://plantuml.com/). This makes it really easy to create the diagrams and edit them quickly, and there's a plug-in available for VS code. The downside is that you have to install the Java runtime and GraphViz.

Here's an example of what this looks like in VS Code:

PlantUML Example

### CODE REVIEW

**Comment by johnstairs** [comment]
Nice idea. Have you looked at Mermaid? It's pretty nice:

[](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5cdEFsaWNlLT4-K0pvaG46IEhlbGxvIEpvaG4sIGhvdyBhcmUgeW91P1xuXHRBbGljZS0-PitKb2huOiBKb2huLCBjYW4geW91IGhlYXIgbWU_XG5cdEpvaG4tLT4-LUFsaWNlOiBIaSBBbGljZSwgSSBjYW4gaGVhciB5b3UhXG5cdEpvaG4tLT4-LUFsaWNlOiBJIGZlZWwgZ3JlYXQhXG5cdFx0XHRcdFx0IiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)

**Comment by hansenms** [comment]
This is awesome, we have a few requests for diagrams like this. 

@YazanMSFT have you considered mermaid graph instead: https://mermaid-js.github.io/ 

There is also a vscode extension (https://marketplace.visualstudio.com/items?itemName=vstirbu.vscode-mermaid-preview) and it is all javascript, so no additional tool installation required. 

It is also easy to inline the mermaid graphs into markdown pages. @johnstairs has done some of that.

**Comment by namadabu** [comment]
This is such a helpful idea. Thanks for doing this!

**Comment by namadabu** [comment]
Sorry closed it by mistake!

**Comment by YazanMSFT** [comment]
> Nice idea. Have you looked at Mermaid? It's pretty nice:
> 
> [](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5cdEFsaWNlLT4-K0pvaG46IEhlbGxvIEpvaG4sIGhvdyBhcmUgeW91P1xuXHRBbGljZS0-PitKb2huOiBKb2huLCBjYW4geW91IGhlYXIgbWU_XG5cdEpvaG4tLT4-LUFsaWNlOiBIaSBBbGljZSwgSSBjYW4gaGVhciB5b3UhXG5cdEpvaG4tLT4-LUFsaWNlOiBJIGZlZWwgZ3JlYXQhXG5cdFx0XHRcdFx0IiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)

I did not, but that looks better. I'll switch to that, thanks for the suggestion!

**Comment by YazanMSFT** [comment]
> This is awesome, we have a few requests for diagrams like this.
> 
> @YazanMSFT have you considered mermaid graph instead: https://mermaid-js.github.io/
> 
> There is also a vscode extension (https://marketplace.visualstudio.com/items?itemName=vstirbu.vscode-mermaid-preview) and it is all javascript, so no additional tool installation required.
> 
> It is also easy to inline the mermaid graphs into markdown pages. @johnstairs has done some of that.

Didn't know that we used it before :) I'll switch to that. Thanks!

File: docs/flow diagrams/create-resource.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    participant AuthorizationService
+    participant ResourceReferenceResolver
+    participant ResourceWrapperFactory
+    participant SearchIndexer
+    participant FhirDataStore
+    Client->>FhirServer: POST Resource
+    FhirServer->>Middleware:
+    Middleware->>FhirController:
+    FhirController->>Mediatr: Create Resource
+    Mediatr->>CreateResourceHandler: Resource
```
Comment (namadabu) [inline docs/flow diagrams/create-resource.md]: nit: Can this be `CreateResourceRequest` to match with the `UpsertResourceResponse` that is returned to the `Mediatr` below ?

File: docs/flow diagrams/search-api.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    Mediatr->>SearchResourceHandler: SearchResourceRequest
+    SearchResourceHandler->>AuthorizationService: CheckAccess
+    AuthorizationService->>SearchResourceHandler:
+    SearchResourceHandler->>SearchService: Search by ResourceType, queries
+    SearchService->>SearchOptionsFactory: Create SearchOptions
+    SearchOptionsFactory->>SearchService:
+    SearchService->>SearchInternalAsync: Data store dependent
+    SearchInternalAsync->>SearchService: SearchResult
+    SearchService->>SearchResourceHandler: SearchResult
+    SearchResourceHandler->>Mediatr: SearchResourceResponse
```
Comment (namadabu) [inline docs/flow diagrams/search-api.md]: Would it be useful to include a step to call out that we return a bundle within the `SearchResourceResponse`?  Something like 
` SearchResourceHandler->>BundleFactory: SearchResult`
` BundleFactory->>SearchResourceHandler: Bundle`

File: docs/flow diagrams/search-sql-server-md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+```mermaid
```
Comment (namadabu) [inline docs/flow diagrams/search-sql-server-md]: typo in filename extension: `.md`

File: docs/flow diagrams/create-resource.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+```mermaid
+sequenceDiagram
+    participant Client
+    participant FhirServer
```
Comment (namadabu) [inline docs/flow diagrams/create-resource.md]: I haven't used mermaid before. So just curious why this file has `participant` values whereas the others don't have it. The VS Code previews did not seem to show any difference.

Comment (YazanMSFT) [reply docs/flow diagrams/create-resource.md]: It isn't required. I haven't used it before either, when I started the first example in the documentation had them so I put them in, then later I saw an example without it. Forgot to clean up the first one, thanks for catching it!

**Pull request merged**